Create location based posts

KronosDP opened the pull request:
This pull request introduces several significant changes to the `pom.xml` file and adds new functionality related to handling geographical locations and posts in the project. The most important changes include renaming the artifact, updating dependencies, and adding new classes and interfaces to support location-based queries and services.

### Changes to `pom.xml`:

* Renamed the artifact and updated the project name and description to `authentication`.
* Updated the JUnit dependency to `junit-jupiter` version `5.11.2`.
* Added new dependencies for PostgreSQL, DataFaker, Spring Data JPA, Spring Security, H2 Database, and others. [[1]](diffhunk://#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R82-R114) [[2]](diffhunk://#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L97-R174)
* Added the `jacoco-maven-plugin` for code coverage reporting.

### New Classes and Interfaces:

* [`GeometryConfiguration`](diffhunk://#diff-97da0352d3618c28c63046c9592ab6316b079561f3f313a9701957b11a7897deR1-R16): Provides a `GeometryFactory` bean for creating geometric shapes.
* [`PostController`](diffhunk://#diff-ccd762a2360598b200c84f0e103d2dd1711b47406529a58ec8d5d224c9672828R1-R104): Handles HTTP requests for posts, including location-based queries.
* [`Location`](diffhunk://#diff-950aab07c9119af885262b9146d9356e11ddab3df83222806d2caa5229041192R1-R16): Represents a geographic location with latitude and longitude.
* [`Post`](diffhunk://#diff-9b99a34203e0311544124469b15573cecf54ba129cb2d719a141c087d20f75ddR1-R138): Represents a post entity with location information.
* [`UserLocation`](diffhunk://#diff-3b0777dae53cd4d05dc52457d910c3ef982f3c0296e6a82d9aa950b8bf9fa6bcR1-R34): Represents the geographical location of a user.
* [`PostRepository`](diffhunk://#diff-9e44453c3a0bdc1b25d2e2aa02be6002d171ef9275dc7d384a22c0a7f6478effR1-R41): Defines methods for querying posts by location and other criteria.
* `LocationProvider` and `LocationService`: Interfaces for providing and managing location data. 
* [`LocationServiceImpl`](diffhunk://#diff-a36a474fc4b48ebff7b322bd976ed92b77a10677337a88e940443b8f69528449R1-R42): Implements the `LocationService` interface.
* [`PostService`](diffhunk://#diff-0d5c44762ee8c9713817825d852a9227791773f2653e41cd93b2bb4adb7d74b6R1-R56): Defines methods for finding and querying posts based on location and other filters.

Frmeta reviewed (changes requested):
Thank you @KronosDP for the pull request, however I don't think it's ready yet for merge to main. Here's my consideration:
- Post id datatype should be UUID instead of long
- Post `content` should be renamed to `caption` (to make it consistent with drawio)
- Don't use category, use tags instead to support multiple tags (like hashtags in social media)
- Variable postedBy that is foregin key that points toward User Id with type UUID
- Post & Comment should inherit abstract class called BasePost (more info in drawio I've sent in discord server)
- boolean isPremium/isVerified to distinguish premium & nonpremium posts
- image link to keep link to image (since image hosting might be troublesome, we can use dummy link first for debugging)
- I think Location.java & UserLocation.java both serve no purpose and should be discarded
I attached related documents below:

![Screenshot 2025-03-03 174346](https://github.com/user-attachments/assets/8ae22bad-c488-4aaa-9222-1cc24c565fc7)
Link to drawio ERD : https://app.diagrams.net/#G1i_eD7w5hR0r3lAsSTQ57L4Qxi8_pAmBJ

KronosDP commented:
Thank you for your suggestion @Frmeta, I will try address your comments one by one. 

> Post id datatype should be UUID instead of long

I have refactored the datatype to use UUID. This is rather an easy fix because I used Intellij to do the datatype migration.

> Post content should be renamed to caption (to make it consistent with drawio)

I have also fixed this one using Intellij, 

> Don't use category, use tags instead to support multiple tags (like hashtags in social media)

As a Product Owner I think that we need to make this to be just categories that are made by use. I don't recommend the users to be able to make their own tags because I think that the app will be too free and people will come up with all sort of hastags. We want people to behave in this app because that what makes us different than our competitior (like twitter or insta). 

> Variable postedBy that is foregin key that points toward User Id with type UUID

I think this issue can be fixed when the authentication microservice is done.

> Post & Comment should inherit abstract class called BasePost (more info in drawio I've sent in discord server)

This is rather interesting. I actually want to discuss about the existance of `GeometryFactory` class that exists from `org.locationtech.jts.geom`. I think we can use that to make langitude and longitude field of the Post. So, actually the Post that I've been making is actually the Abstract class. We can make the extension class this week. 

> image link to keep link to image (since image hosting might be troublesome, we can use dummy link first for debugging)

Just like the above comment, I think that this can be implemented this week. Thank you for the input.

> boolean isPremium/isVerified to distinguish premium & nonpremium posts

This is an easy fix to implement. But I think that we should implement this when we are working on premium forum feature.

> I think Location.java & UserLocation.java both serve no purpose and should be discarded

Yes, it's true, that is just a giberish from the AI generated boilerplate. Thank you for pointing it out.

Frmeta commented:
This PR is a well-structured and thoughtful refactor that enhances the maintainability, readability, and efficiency of the codebase. Overall, this PR significantly improves code quality, structure, and maintainability while keeping the application robust and performant. Great job!
